src/helpers/visualisation.py

```diff
@@ -428,4 +428,715 @@ def plot_differences_for_group(
+    turn_col: str,
+    label_col: str,
+    weight_col: Optional[str] = None,           # None: count rows; else sum weights
+    top_k_labels: Optional[int] = None,         # collapse others to "Other"
+    exclude_labels: Optional[Sequence[str]] = None,
+    turn_cutoff: Optional[int] = None,          
+    labels_order: Optional[Sequence[str]] = None,
+    palette: Optional[Sequence] = None,
+    title: str = "",
+    xlabel: str = "User Turn Index",
+    ylabel: str = "Proportion",
+    col_wrap: int = 3,
+    height: float = 3.0,
+    aspect: float = 1.3,
+    xticks_step: Optional[int] = None,          # e.g., 2 for even-turn spacing labels
+    dpi: int = 150,
+    output_path: Optional[str] = None,
+    show: bool = True, 
+) -> plt.Figure:
+    data = df.copy()
+
+    if exclude_labels:
+        data = data[~data[label_col].isin(exclude_labels)]
+
+    if top_k_labels is not None:
+        counts = data[label_col].value_counts()
+        keep = set(counts.index[:top_k_labels])
+        data[label_col] = np.where(data[label_col].isin(keep), data[label_col], "Other")
+
+    if turn_cutoff is not None:
+        data = data[data[turn_col] < turn_cutoff]
+
+    # compute counts/proportions per (group, turn, label)
+    if weight_col is None:
+        grp = (
+            data.groupby([group_col, turn_col, label_col])
+                .size().rename("count").reset_index()
+        )
+    else:
+        grp = (
+            data.groupby([group_col, turn_col, label_col])[weight_col]
+                .sum().rename("count").reset_index()
+        )
+    grp["total"] = grp.groupby([group_col, turn_col])["count"].transform("sum")
+    grp["proportion"] = grp["count"] / grp["total"]
+
+    # label order + palette
+    labels_sorted = labels_order or sorted(grp[label_col].unique())
+    pal = palette or sns.color_palette("tab10", n_colors=len(labels_sorted))
+
+    g = sns.FacetGrid(
+        data=grp, col=group_col, col_wrap=col_wrap, height=height, aspect=aspect,
+        sharey=True, sharex=True, hue=label_col, palette=pal, hue_order=labels_sorted
+    )
+    g.map_dataframe(sns.lineplot, x=turn_col, y="proportion", marker="o")
+    g.add_legend(title="Label", bbox_to_anchor=(1.02, 0.5), loc="center left")
+    g.set_axis_labels(xlabel, ylabel)
+    g.set_titles("{col_name}")
+
+    # x ticks
+    if turn_cutoff is not None or xticks_step is not None:
+        xmax = int(grp[turn_col].max()) if turn_cutoff is None else turn_cutoff - 1
+        ticks = range(0, xmax + 1, xticks_step or 1)
+        for ax in g.axes.flatten():
+            ax.set_xticks(ticks)
+
+    plt.subplots_adjust(bottom=0.15, right=0.82, top=0.9)
+    g.fig.suptitle(title, y=0.98, fontsize=13)
+
+    if output_path:
+        g.fig.savefig(output_path, bbox_inches="tight", dpi=dpi)
+    if show:
+        plt.show()
+    else:
+        plt.close(g.fig)
+    return g.fig
+
+
+# Overall label trends over user turns (single axes)
+def plot_label_trends_overall(
```

Comment:
more specific function name. This sounds quite generic. Use the name of the type of chart.